base/combinatorics.jl

```diff
@@ -232,7 +233,7 @@ julia> A
  1
 ```
 """
-invpermute!(a, p::AbstractVector) = invpermute!!(a, copymutable(p))
+invpermute!(v, p::AbstractVector) = copyto!(v, v[invperm(p)])
```

Comment:
Wouldn't it be better to use:
```jl
w = similar(v)
w[p] = v
return w
```
?

Reply:
Unfortunately, that violates `invpermute!(v, p) === v`.

This is a candidate implementation of `invpermute`, a replacement of `v[invperm(p)]` that runs 2-3x faster, but we'd need to demonstrate a convincing use case to include it in base rather than Permutations.jl.

Reply:
Sorry, the last line should be `return copyto!(v, w)`

Reply:
On my machine with a random length-1000 permutation and `v::Vector{Float64}`, my version is 30% faster as well as allocating less memory.

Reply:
Or even simpler, 
```jl
invpermute!(v, p::AbstractVector) = (x[p] = x; x)
```

Reply:
And again, I think that we should take out invpermute! in 2.0.

Reply:
While these changes will make the choice to have `invpermute!` in base look silly, given that we do have `invpermute!` in 1.x, I think this version is better than what we have right now because it is faster in most cases and consequently better facilitates migration away from `Base.invpermute!!`

```julia
julia> @benchmark invpermute!(v, p) setup=(v=rand(3000); p=shuffle(1:3000)) evals=1 # This PR
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):   4.647 μs …  13.241 ms  ┊ GC (min … max):  0.00% … 99.75%
 Time  (median):      6.744 μs               ┊ GC (median):     0.00%
 Time  (mean ± σ):   18.594 μs ± 265.536 μs  ┊ GC (mean ± σ):  26.63% ±  2.00%

  ▇█▅▃▃▂▁▁▁▃▃▂▁                                                ▂
  ███████████████▇▇▆▆▆▆▅▅▆▅▅▆▄▄▅▅▄▅▃▄▅▅▄▄▃▃▄▅▄▁▃▁▁▃▃▅▄▃▄▁▃▁▃▁▃ █
  4.65 μs       Histogram: log(frequency) by time      90.6 μs <

 Memory estimate: 23.48 KiB, allocs estimate: 2.

julia> @benchmark invpermute!(v, p) setup=(v=rand(3000); p=shuffle(1:3000)) evals=1 # 1.7.2
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  15.289 μs …  16.599 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     20.938 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   45.645 μs ± 302.911 μs  ┊ GC (mean ± σ):  9.46% ± 1.99%

  █▄▆▆▃▂▁▁▁▁                                                   ▂
  ████████████▇█▇▆▆▆▆▆▅▆▅▆▅▆▄▄▄▆▄▄▄▃▃▁▄▃▄▄▄▄▄▄▄▅▁▃▁▄▄▃▁▃▄▁▁▃▄▄ █
  15.3 μs       Histogram: log(frequency) by time       377 μs <

 Memory estimate: 23.48 KiB, allocs estimate: 2.

julia> @benchmark Base.invpermute!!(v, p) setup=(v=rand(3000); p=shuffle(1:3000)) evals=1 # This PR
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):   9.758 μs …   8.325 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     12.275 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   20.373 μs ± 137.556 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

  ▇█▆▃▂▁  ▁▁                                                   ▂
  ███████████▇▇▇▆▅▇▆▅▅▆▅▅▆▆▆▅▅▅▆▅▃▄▅▆▅▄▅▅▄▄▄▄▃▄▄▃▄▃▁▁▃▄▃▁▅▄▁▁▄ █
  9.76 μs       Histogram: log(frequency) by time       111 μs <

 Memory estimate: 0 bytes, allocs estimate: 0.
```